### PR TITLE
Clear pause state on step action

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #4024 

### Summary of Changes

* will  also help address  some of the intermittents we see w/ wasm and other tests by removing the chance for unsafe calls

### Tests

* unit tests

### CODE REVIEW

File: src/reducers/pause.js [hunk 1]
```diff
@@ -80,12 +87,7 @@ function update(state: PauseState = State(), action: Action): PauseState {
         frameScopes: { ...state.frameScopes, [selectedFrameId]: scopes }
       };
     case "RESUME":
-      return Object.assign({}, state, {
-        pause: null,
-        frames: null,
-        selectedFrameId: null,
-        loadedObjects: {}
-      });
+      return { ...state, emptyPauseState };
```
Comment (jimblandy) [inline src/reducers/pause.js]: Doesn't this create an object that looks like state, but with a new property named `emptyPauseState`?

File: src/actions/pause/commands.js [hunk 2]
```diff
@@ -18,12 +19,11 @@ type CommandType = "stepOver" | "stepIn" | "stepOut" | "resume";
...
-    dispatch({ type: "COMMAND", value: { type } });
-
-    await client[type]();
-
-    dispatch({ type: "CLEAR_COMMAND" });
+    return dispatch({
+      type: "COMMAND",
+      command: type,
+      [PROMISE]: client[type]()
+    });
```
Comment (jimblandy) [inline src/actions/pause/commands.js]: Is this neglecting to put a `status` property on this action, or is this not a COMMAND `PauseAction`?

Comment (jimblandy) [reply src/actions/pause/commands.js]: I see; this communicates with the promise middleware.

File: src/reducers/pause.js [hunk 3]
```diff
@@ -153,10 +155,9 @@ function update(state: PauseState = State(), action: Action): PauseState {
       });
 
     case "COMMAND":
-      return { ...state, command: action.value.type };
-
-    case "CLEAR_COMMAND":
-      return { ...state, command: "" };
+      return action.status === "start"
+        ? { ...state, ...emptyPauseState, command: action.command }
+        : { ...state, command: "" };
```
[the remark this answers is not in the document]
Comment (jimblandy) [reply src/reducers/pause.js]: So what this is really doing is replacing the `COMMAND` / `CLEAR_COMMAND` pair with a single `COMMAND` action, and using the `AsyncStatus` attached by the promise middleware to discriminate between the two behaviors.

File: src/reducers/pause.js [hunk 4]
```diff
@@ -165,7 +166,7 @@ function update(state: PauseState = State(), action: Action): PauseState {
       };
 
     case "NAVIGATE":
-      return { ...state, debuggeeUrl: action.url };
+      return { ...state, ...emptyPauseState, debuggeeUrl: action.url };
```
Comment (jimblandy) [inline src/reducers/pause.js]: This seems like an unrelated change. Is it somehow necessitated by the changes to the `COMMAND` action?

File: src/test/mochitest/head.js [hunk 5]
```diff
@@ -343,13 +344,22 @@ function isPaused(dbg) {
...
  */
 async function waitForPaused(dbg) {
-  // We want to make sure that we get both a real paused event and
-  // that the state is fully populated. The client may do some more
-  // work (call other client methods) before populating the state.
-  let loading = waitForDispatch(dbg, "LOAD_OBJECT_PROPERTIES");
-  await waitForThreadEvents(dbg, "paused");
-  await waitForState(dbg, state => isPaused(dbg));
-  await loading;
+  function scopeIsLoaded(dbg) {
```
Comment (jimblandy) [inline src/test/mochitest/head.js]: Rather than digging around in the state and counting properties, would it be possible for, say, something in `src/reducers` to provide a predicate on a state that indicates whether the scopes are fully loaded? It seems to me that duplicating the knowledge of that structure's details here is asking for trouble.

File: src/test/mochitest/head.js [hunk 6]
```diff
@@ -343,13 +344,22 @@ function isPaused(dbg) {
...
+    if (loadedObjects.length == 0) {
+      return false;
+    }
+
+    return Object.keys(loadedObjects[0]).length !== 0;;
+  }
+
+  await waitForState(
+    dbg,
+    state => isPaused(dbg) && scopeIsLoaded(dbg),
```
Comment (jimblandy) [inline src/test/mochitest/head.js]: Is the `state` being passed to this function the same value retrieved by `dbg.getState()`? It seems like it would make sense for `scopeIsLoaded` to be a predicate on states.

File: src/test/mochitest/head.js [hunk 7]
```diff
@@ -679,9 +688,9 @@ function togglePauseOnExceptions(
  * @return {Promise}
  * @static
  */
-function invokeInTab(fnc) {
+async function invokeInTab(fnc) {
   info(`Invoking function ${fnc} in tab`);
-  return ContentTask.spawn(gBrowser.selectedBrowser, fnc, function*(fnc) {
+  ContentTask.spawn(gBrowser.selectedBrowser, fnc, function*(fnc) {
```
Comment (jimblandy) [inline src/test/mochitest/head.js]: Won't failing to return the promise here break `uncaughtException` and `caughtException` in `src/test/mochitest/browser_dbg-pause-exceptions.js`?

File: src/test/mochitest/browser_dbg-sourcemaps.js [hunk 8]
```diff
@@ -78,13 +78,14 @@ add_task(async function() {
   assertPausedLocation(dbg);
 
   await stepIn(dbg);
-  // await waitForMappedScopes(dbg);
+  await waitForSelectedSource(dbg, "times2.js")
```
Comment (jimblandy) [inline src/test/mochitest/browser_dbg-sourcemaps.js]: Is this somehow related to the change to the way the pause state is managed? If not, it should be a separate pull request.

File: src/reducers/pause.js (hunk 1 above)
Comment (jasonLaster) [reply src/reducers/pause.js]: Yep, good catch. Turns out that we can drop this entirely now that we no longer update the state when we are resumed

File: src/reducers/pause.js (hunk 4 above)
Comment (jasonLaster) [reply src/reducers/pause.js]: It is something that happened before, but this change makes it more explicit. It is helpful to do it here to prevent future regressions. Also simpler to test

File: src/test/mochitest/head.js (hunk 6 above)
Comment (jasonLaster) [reply src/test/mochitest/head.js]: 👍

File: src/test/mochitest/head.js [hunk 9]
```diff
@@ -343,13 +344,13 @@ function isPaused(dbg) {
...
-  // work (call other client methods) before populating the state.
-  let loading = waitForDispatch(dbg, "LOAD_OBJECT_PROPERTIES");
-  await waitForThreadEvents(dbg, "paused");
-  await waitForState(dbg, state => isPaused(dbg));
-  await loading;
+  const { getSelectedScope, hasLoadingObjects } = dbg.selectors;
+
+  return waitForState(
+    dbg,
+    state => isPaused(dbg) && getSelectedScope(state) && !hasLoadingObjects(state),
```
Comment (jasonLaster) [inline src/test/mochitest/head.js]: we're now checking to see if we have a fetched scope and that there aren't any pending object fetches which is pretty nice

File: src/test/mochitest/head.js (hunk 7 above)
Comment (jasonLaster) [reply src/test/mochitest/head.js]: we aren't using the promise, but i shouldn't have removed it either

File: src/reducers/pause.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasonLaster) [reply src/reducers/pause.js]: yep

**Comment by codehag** [comment]
This seems to have been split into smaller prs, should it be closed?
